app/root.tsx

```diff
@@ -49,6 +50,7 @@ export const action = async ({ request }: ActionArgs) => {
   const formData = await request.formData();
   const originalName = formData.get('originalName');
   if (typeof originalName === 'string') {
+    logger.info(`Admin (${originalName}) stopped impersonation`);
```

Comment:
Should this not be done in setEffectiveUsername (add an else part to that condition)?

Reply:
I tried that, but we seem to go through that code path on normal startup (e.g., login), and there's no way to differentiate a login from stopping impersonation.  It meant that it always logged that the admin user was stopping impersonation on regular login.  I used this method as a way to indicate it's really done.  I could add a function to do it and log, but it would be the same code path as this.

Reply:
My issue is code maintainability. I'm afraid that this current solution will break in the next refactor / if that fn gets called from an other place


What would you think about changing the signature of this:
```TS
export async function setEffectiveUsername(
  username: string,
  effectiveUsername: string | null
): Promise<string> {
```

to somehing like this overloaded signature (have not tested this lang element yet, but according to the docs it should work :) )
```ts
function setImpersonation(action: 'init', username: string): Promise<string>;
function setImpersonation(action: 'start', username: string, effectiveUsername: string): Promise<string>;
function setImpersonation(action: 'end', username: string): Promise<string> {
   ...
}
```

Reply:
that way you could hook onto the action parameter, and push the proper log

Reply:
Maybe we should add `effectiveUsername` in the log message here, to make it easier to track in the log.
For example, perhaps an admin decides to do multiple impersonations, and we need to go through a lot of log messages to figure out which is which.